optuna_dashboard/static/components/GraphContour.tsx

```diff
@@ -0,0 +1,259 @@
+  }
+
+  const handleYAxisChange = (e: ChangeEvent<{ value: unknown }>) => {
+    setYAxis(e.target.value as string)
+  }
+
+  useEffect(() => {
+    if (study !== null) {
+      plotContour(study, objectiveId, xAxis, yAxis)
+    }
+  }, [study, objectiveId, xAxis, yAxis])
+
+  return (
+    <Grid container direction="row">
+      <Grid item xs={3}>
+        <Grid container direction="column">
+          {study !== null && study.directions.length !== 1 ? (
+            <FormControl component="fieldset" className={classes.formControl}>
+              <FormLabel component="legend">Objective ID:</FormLabel>
+              <Select value={objectiveId} onChange={handleObjectiveChange}>
+                {study.directions.map((d, i) => (
+                  <MenuItem value={i} key={i}>
+                    {i}
+                  </MenuItem>
+                ))}
+              </Select>
+            </FormControl>
+          ) : null}
+          <FormControl component="fieldset" className={classes.formControl}>
+            <InputLabel id="parameter1">X Axis Parameter</InputLabel>
+            <Select value={xAxis} onChange={handleXAxisChange}>
+              {paramnames.map((x) => (
+                <MenuItem value={x} key={x}>
+                  {x}
+                </MenuItem>
+              ))}
+            </Select>
+          </FormControl>
+          <FormControl component="fieldset" className={classes.formControl}>
+            <InputLabel id="parameter2">Y Axis Parameter</InputLabel>
+            <Select value={yAxis} onChange={handleYAxisChange}>
+              {paramnames.map((x) => (
+                <MenuItem value={x} key={x}>
+                  {x}
+                </MenuItem>
+              ))}
+            </Select>
+          </FormControl>
+        </Grid>
+      </Grid>
+      <Grid item xs={6}>
+        <div id={plotDomId} />
+      </Grid>
+    </Grid>
+  )
+}
+
+const plotContour = (
+  study: StudyDetail,
+  objectiveId: number,
+  xAxis: string,
+  yAxis: string
+) => {
+  if (document.getElementById(plotDomId) === null) {
+    return
+  }
+
+  const layout: Partial<plotly.Layout> = {
+    title: "Contour",
+    margin: {
+      l: 50,
+      r: 50,
+    },
+    xaxis: {
+      gridcolor: "#f2f5fa",
+      gridwidth: 1,
+      zerolinecolor: "#f2f5fa",
+      zerolinewidth: 1.5,
+    },
+    yaxis: {
+      gridcolor: "#f2f5fa",
+      gridwidth: 1,
+      zerolinecolor: "#f2f5fa",
+      zerolinewidth: 1.5,
+    },
+    plot_bgcolor: "#E5ecf6",
+  }
+
+  const trials: Trial[] = study !== null ? study.trials : []
+
+  if (trials.length === 0) {
+    plotly.react(plotDomId, [], layout)
+  }
+
+  const filteredTrials = trials.filter(
+    (t) => t.state === "Complete" || t.state === "Pruned"
+  )
+
+  let paramNames = new Set<string>(trials[0].params.map((p) => p.name))
+  filteredTrials.forEach((t) => {
+    paramNames = new Set<string>(
+      t.params.filter((p) => paramNames.has(p.name)).map((p) => p.name)
+    )
+  })
```

Comment:
These lines are duplicated to L27-L37. I think we can reduce these computation costs. Could you refactor this?

Reply:
I tried to remove this part (since it was duplicate) but it was giving an error. However, I'll try to refactor this.